pycaw/magic.py

```diff
@@ -0,0 +1,812 @@
+        back to the super().__init(new_magic_root_session) -
+
+        the new_magic_root_session it is transferred by setting
+        a class attribute and picking it up in the MagicSession.__init__()
+
+        class PassingViaArg(MagicSession):
+            def __init__(self, new_magic_root_session, *args, **kwargs):
+                super().__init__(new_magic_root_session)
+
+        class PassingViaClassAttr(MagicSession):
+            def __init__(self, *args, **kwargs):
+                super().__init__()
+        """
+        cls._passed_magic_root_session = new_magic_root_session
+
+        magic_session = cls(*args, **kwargs)
+        return magic_session
+
+    def __str__(self):
+        return (f"<{self.__class__.__name__} "
+                f"app_exe='{self.magic_root_session.app_exe}'/>")
+
+    # easy control:
+    # TODO: handle incorrect input or raise exception.
+    # also handle failing com calls
+    # (failing in terms of the retrieved value is not 'S_OK')
+    @property
+    def state(self):
+        return self.magic_root_session.state
+
+    @property
+    def volume(self):
+        return self.magic_root_session.volume
+
+    @volume.setter
+    def volume(self, volume):
+        self.magic_root_session._sav.SetMasterVolume(volume, self.guid)
+
+    @property
+    def mute(self):
+        return self.magic_root_session.mute
+
+    @mute.setter
+    def mute(self, mute):
+        self.magic_root_session._sav.SetMute(mute, self.guid)
+
+    def toggle_mute(self):
+        new_mute = not self.mute
+        self.mute = new_mute
+        return new_mute
+
+    def step_volume(self, step=0.1):
+        current = self.volume
+        new = max(0, min(1, current + step))
+        if new != current:
+            self.volume = new
+            return new
+        return None
+
+
+class MagicRootSession(COMObject):
+    """Base session control with callback functionality"""
+
+    _com_interfaces_ = (IAudioSessionEvents,)
+
+    # ======= DECODE RETURNED INT VALUE =======
+    # see audiosessiontypes.h
+    AudioSessionState = (
+        "Inactive",
+        "Active",
+        "Expired"
+    )
+
+    def __init__(self, audio_session_control2, iid, magic_manager):
+        self._ctl2 = audio_session_control2
+        self.app_exe = self._get_app_exe()
+        self._sav = None
+        self.magic_manager = magic_manager
+        self.iid = iid
+
+        self.magic_app = None
+        self.magic_session = None
+
+        self.volume = None
+        self.mute = None
+
+        self.state_id = self._ctl2.GetState()
+        new_state = self.AudioSessionState[self.state_id]
+        self.state = (new_state, self.state_id)
+
+        self._activated = False
+
+        self.register_notification()
+
+        log.info(f":: created {self}")
+
+    # XXX remove old session:
+    # doesnt work this way either...
+    #
+    # def IUnknown_Release(self, *args, **kwargs):
```

Comment:
I'd rather ditch any commented out code from the code base if possible

Reply:
https://github.com/AndreMiras/pycaw/pull/42#issuecomment-841659891